% using 'oldplainarticle' for one-col draft, see comment below 'NOTE DOCCLASS'
% comment this line out (and others) and replace w/ IEEEtran
\documentclass[fleqn,10pt]{olplainarticle}


% using 'IEEEtran' for a two-col draft, closer to expected journal, see comments below 'NOTE DOCCLASS'
% comment this line out (and others) and replace w/ oldplainarticel
%\documentclass[journal,draftcls]{IEEEtran}
% SEE https://ctan.mirrors.hoobly.com/macros/latex/contrib/IEEEtran/IEEEtran_HOWTO.pdf

\frenchspacing
\usepackage{xr-hyper}
\usepackage{hyperref}                   % links
\usepackage{xcolor}
\usepackage{amsmath}                    % math symbols
\usepackage{lineno}
\usepackage{multirow}
%\usepackage[ruled,vlined,noend]{algorithm2e}  % pseudo-code; NOTE do not use float env provided by package w/ IEEE
%\usepackage{xargs}                      % Use more than one optional parameter in a new commands
%\usepackage{todonotes}                  % notes in sidebar
%\usepackage{authblk}                    % support footnote style author / affiliation
%\usepackage[htt]{hyphenat}              % used to allow linebreaks for teletype (\texttt)
%\usepackage[acronym]{glossaries}        % acronymns: \newacronym
%\loadglsentries{glossary}

\hypersetup{urlcolor=cyan,colorlinks=true,linkcolor=blue}
\usepackage{array}
\usepackage{booktabs}
\usepackage{tabularx}


\makeatletter
%\newcommand*{\addFileDependency}[1]{% argument=file name and extension
%  \typeout{(#1)}
%  \@addtofilelist{#1}
%  \IfFileExists{#1}{}{\typeout{No file #1.}}
%}
%\makeatother

%\newcommand*{\myexternaldocument}[1]{%
%    \externaldocument{#1}%
%    \addFileDependency{#1.tex}%
%    \addFileDependency{#1.aux}%
%}
%\myexternaldocument{Supplement}

\usepackage{setspace}
\doublespacing  % NOTE comment out to use single spacing


%\RequirePackage{listings}


% NOTE on bibliography, when working offline, need to manually copy the Mendeley bibfile at 
% ~/Documents/BibFiles/StrawberryDomestication_Manuscript.bib to this folder as references.bib 
% if you want to update the bibliography.
% Online Mendeley syncs to the correct folder and file on Overleaf


\linenumbers

\begin{document}
%\newacronym{te}{TE}{Transposable Element}

% NOTE DOCCLASS used w/ oldplainarticle
\keywords{Transposable Elements, Genomics, Genome Evolution, Bioinformatics, Python}
% NOTE DOCCLASS used w/ IEEE
%\begin{IEEEkeywords}
%    Transposable Elements, TE, Syntelog
% \end{IEEEkeywords}


% Use option lineno for line numbers 
\title{The Phenotypic Consequences of Transposable Elements in Cultivated Strawberry}

\author[1,2]{Scott J. Teresi}
\author[1]{Jordan R. Brock}
\author[1]{Ning Jiang}
\author[1,2]{Patrick P. Edger}
\affil[1]{Department of Horticulture, Michigan State University, East Lansing, MI}
\affil[2]{Genetics and Genome Sciences Program, Michigan State University, East Lansing, MI}

\begin{abstract}
\textbf{Background:}
% The cultivated strawberry (\textit{Fragaria x ananassa}), unlike most crops, is unique because it is a recent domesticate, formed by the interspecific hybridization of wild octoploid species.
% Here, armed with the availability of a wide-array of genomic and phylogentic resources that include wild diploid, wild octploid, and a high-quality reference genome for \textit{Fragaria x ananassa} we examined the genome for signatures of domestication.
% While previous research has shown that the dominant subgenome possesses the least amount of transposable elements (TEs) and that they are important contributors to genotypic and phenotypic variation, the potential influence of TEs on the domestication of strawberry has been relatively understudied.
% %Transposable elements (TEs) are major forces of genetic mutation and phenotypic diversity.
% %A wide-range of traits in a variety of crop species have been shown to be influenced by transposable elements.
% %This is driven by the capacity of TEs to act as alternative promoters, shuffle gene sequences, or influence methylation and chromatin accessibility dynamics near genes.
% \newline
% \textbf{Results:}
% Here we constructed a TE-pangenome for strawberry and show that although the relative TE content between the wild and domesticated strawberry genomes is quite similar, there are large TE differences around genes that suggest their role in the domestication process.
% We show large transposon dynamics around positionally conserved genes suggesting rapid evolution.
% Additionally, we show that a number of defense, biosynthesis, and development genes are enriched for transposon presence and have high expression, suggesting the co-opting or domestication of TEs during the domestication of strawberry. % (TODO verify this!!!)
% \newline
% \textbf{Conclusions:}
% Together, this study describes the TE-gene dynamics of strawberry, showing the potential influence of TEs, and suggests a ...
\end{abstract}

\flushbottom
\maketitle
\thispagestyle{empty}

\newpage

% TODO talk about pangenomes in introduction, talk about core and dispensable genes and how there are a lot of dispensable genes
% TODO talk about lack of shared strawberry genes in intro
% TODO talk about TEs being bad and then talk about how they can be good
% TODO talk about mechanisms or patterns to get rid of TEs and why it is exceptional for a gene to have a ton of TE near it.


%Furthermore Hardigan et al. \cite{Hardigan2020} suggests that ``interspecific complementation, a broader pool of potentially adaptive alleles, and the masking of deleterious mutations could be more important than fixed heterosis'' in strawberry.\\

% Strawberry is a recent domesticate.
% Strawberry is really diverse.
% Domestication processes in strawberry are not well understood.
% The role of transposable elements in domestication in general is not well understood either.
% 

\section{Background:}
Cultivated garden strawberry (\textit{Fragaria x ananassa}, Duchesne ex Rozier, \textit{2n} $=$ 8X $=$ 56) is an allo-octoploid species formed from the hybridization of two New World strawberry species \textit{Fragaria virginiana} and \textit{Fragaria chiloensis}.
It is a major fruit crop; 1,211,090 tonnes were produced in the United States in 2021... TODO.
%(\href{TODO CITE FAOSTAT}{https://data.un.org/Data.aspx?d=FAO&f=itemCode\%3A544}) and the US production value was evaluated as \$3,398,943,000 in 2023 (\href{TODO CITE USDA}{https://www.nass.usda.gov/Statistics_by_Subject/result.php?C6E91360-DC42-3C4A-8D5C-1392239A0EAF&sector=CROPS&group=FRUIT%20%26%20TREE%20NUTS&comm=STRAWBERRIES}.
Its initial hybridization occurred in the 18th century in France, making strawberry one of the most recent domesticated crop species.
However, unlike most domesticated crops, strawberry has not undergone thousands of years of breeding, only about 300 years of breeding.
In addition to its short history of breeding, the construction of a chromosome-scale genome assembly was not available until quite recently when it was completed in 2019 by Edger et al \cite{Edger2019}. \\
% Most work up until that point made use of the diploid wild strawberry genome, which is an extant relative to one of the diploid subgenomes of \textit{Fragaria x ananassa} \cite{Shulaev2011,Edger2018b}.


For many years the octoploid nature of the genome presented a large barrier to its study, as the unambiguous assignment of homeologs to each subgenome was rather difficult and complicated by a lack of long-read data and frequent homeologous exchanges among the subgenomes.
Once the technical hurdle of generating a high-quality reference genome was overcome, researchers in short time made great strides in understanding the genome and genetic diversity of strawberry.
% TODO gonna have to edit this
Long thought to be a genetically nondiverse genome, research has shown that cultivated strawberry is actually a highly-diverse genome, as over 70\% of alleles from wild populations have been maintained \cite{Hardigan2020,Hardigan2021}.
Additional work by Feldman et al has shown that ... (something about heterosis and genetic gains) \cite{Feldman2024}.
While amazing progress has been made investigating the genomic changes associated with subgenome dominance, the evolutionary origins of the octoploid genome, the genetic diversity of wild and cultivated octoploid strawberry, the role of transposable elements (TEs) in shaping genotypic and phenotypic diversity in strawberry remains to be addressed in greater detail. \\

TEs are reptitive, (sometimes) mobile DNA sequences that constitute a major component of most plant genomes.
They can be roughly divided into 2 major categories, Class I and Class II which are retrotransposons and DNA transposons, respectively.
Class I elements transpose via an RNA intermediate and can be further divided into LTR (long terminal repeat) and non-LTR elements such as LINE and SINE (long interspersed nuclear element, and short interspersed nuclear element, respetively), with LTR elements being the most common in plants.
Class II elements do not utilize an RNA intermediate and transpose via a ``cut-and-paste'' strategy.
Class II elements are also known for their TIRs (terminal inverted repeats) which are one of their primary distinguishing features. \\


TEs are rich sources of genotypic and phenotypic diversity and are near-ubiquitous throughout the plant and animal kingdom.
They can influence the expression of genes in a multitude of ways, including but not limited to, insertional mutagenesis altering the normal sequence of a gene, acting as cis-regulatory elements, shuffling gene sequences, producing aberrant transcripts via read-through transcription of the TE into a neighboring gene, and influencing methylation and chromatin accessibility dynamics near genes \cite{Lisch2013,Choi2018a}.
TEs are frequently targeted by their host genome with repressive chromatin/methylation marks via RNA-directed DNA methylation (RdDM) in order to prevent their transcription and any genotypic mayhem they may cause.
The spread of these repressive marks via the self-reinforcing nature of RdDM is thought to cause methylation ``spill-over'' into sequences adjacent to the TE, potentially downregulating nearby genes.
Likely for this reason, gene expression has been found to be negatively correlated with the density of methylated TEs, and purifying selection works to remove methylated TEs from gene-rich regions \cite{Hollister2009b}.
Like most mutations, most TEs can be considered deleterious, but there are instances where they can be beneficial.
In plants TEs have been shown to influence a wide-range of traits in a variety of species, such as color variation in orange, grape, apple, disease resistance in pepper, pathogen response in wheat, metabolite and environmental variation in tomato, and branching architecture which was relevant to the domestication of maize.
Their influence on phenotypes in strawberry is not well understood and ripe for investigation given recent developments in transposable element annotation software \cite{Ou2019} (TODO CITE PanEDTA), the recent publication of selective sweep and QTL datasets, and the recent availability of high-quality genome assemblies for the wild diploid, wild octoploid, and cultivated octoploid strawberry genomes. \\


% Now I want to move onwards to talking about TE-gene interactions in strawberry as well as some domestication studies.
There is a growing body of evidence that suggests a potential role for TEs in shaping agriculturally relevant traits in strawberry.
For example, Castijello et. al 2020 found that the fruit color variation in several wild and octoploid strawberry cultivars was impacted by multiple, independent TE insertions in or around a MYB10 transcription factor gene \cite{Castillejo2020}.
Fan et al 2024 found a number of eQTL associated with fruit-expressed genes, and observed that a large proportion of TEs overlapped with structural variants in those eQTL identified regions \cite{Fan2022}.
Feldman showed that the genetic gains from breeding for yield are likely not due to heterosis but rather the accrual of additive genetic variation coupled with chemical application practices, and that this occurred ``despite breeding-associated declines in heterozygosity and nucleotide diversity'' \cite{Feldman2024}.
Hardigan et al showed that substantial allelic diversity from wild populations has been maintained in the cultivated strawberry genome, and that there is a substatial amount of genetic variation potentially including alleles that are deleterious and have not yet been purged by breeding \cite{Hardigan2020}. \\


% Final introduction paragraph; summary/approach + main findings
Here we constructed a TE-pangenome for wild and domesticated strawberry and used that to analyze TE and gene differences.
We hypothesized that TEs may have played a role in the domestication of strawberry, and that the TE landscape around genes could be contributing to domestication-related phenotypes.
Here we show that while the relative TE content between the wild and domesticated strawberry genomes is similar, there are large TE differences around genes that suggest a potential role for TEs in the domestication process.
We show that cultivated (domesticated) strawberry has a more compact genome, with less space between genes, and less TE content around genes.
We hypothesized that in the absence of differential selection on TE-gene relationships, the set of TE-dense genomes from both genomes would share a large proportion of functional enrichments, and to our surprise we found little overlap in the functional enrichments of the TE-dense genes.
We also show that the syntelog pairs between the wild and domesticated strawberry genomes have striking differences in TE Density, with the domesticated strawberry syntelogs having less TE Density for all TE types and measurement categories.
Furthermore, we found that TE-dense genes exhibit reduced rates of ortholog identification and found that deeply conserved genes are less likely to be enriched for nearby TEs, which has major implications for the study of novel genes and gene families in strawberry.
Finally, we identified a number of fruit, defense, transcription factor, and development genes that are enriched for novel TE presence in domesticated strawberry, demonstrating the potential for TEs to influence the domestication phenotypes strawberry, and laying the groundwork for future functional validation work. \\ 




% The contributions of heterosis to genetic gains for yield and other domestication traits is unclear Feldman 2024 (this was copies from the Feldman pre-release)
% Lots of genetic variation, including ``unfavorable incompletely dominant alleles that diminish hybrid performance and have not yet been purged by breeding'' (Feldman pre release)


% It is likely that TEs may have played a role in domestication, aside from corn, ....
%Additionally, some hallmarks of selection have been found via the identification of domestication-related selective sweeps as well as QTL associated with fruit quality traits (TODO CITE HARDIGAN 2021a). \\

% Talk about the tradeoff of managing silencing TEs but they also have the potential to create positive change
% Wrap everyting with our appproach, most of the time unique TE-gene instances are the result of a unique phenotype tat is found and then the TE is found to be the causal mutation, here we are starting by looking for TEs and then proceeding from there.

% These datasets have enabled a wide-array of studies to be performed on the strawberry genome as several research groups have worked on identifying selective-sweep regions, in addition to surveying the genomic diversity, evolutionary history, and structure of the strawberry genome (TODO CITE).
% For some time the genetic variation in \textit{Fragaria x ananassa} was considered to be rather limited (TODO CITE), a notion that has been strongly challenged by recent research (TODO CITE HARDIGAN).
% In fact, Hardigan et al showed that genomic diversity is actually quite high, both in the wild founder and domesticated populations.
% Other hallmarks of domesticated genomes are absent in strawberry such as ... TODO.
% Additionally, the genome shows high degree of karyotypic stability and absence of chromosome restructuring \cite{Hardigan2020, Hardigan2021}.


% While great strides have been made in such a short time in understanding the genome and genetic diversity of strawberry, the success of strawberry as an agricultural crop is not fully understood.
% Its high nucleotide diversity, and limited quantitative evidence of heterosis \cite{Stegmeir2010, Rho2012, Hardigan2020} leads us to suspect that initial breeding efforts may have had more to do with the purification/silencing of deleterious alleles and transposable elements (TEs) rather than the fixation of heterotic loci.
% This is in contrast to other crops such as maize, where heterosis is widely known as major contributor to yield TODO CITE, and whose genome has had more time to undergo purifying selection despite its high TE content.
% 
% 
% Given that cultivated strawberry is a highly-diverse recent domesticate with limited evidence of heterosis, we hypothesized that the relatively understudied component of plant genomes, transposable elements (TEs), could be playing an active role in the domestication of strawberry.
% 
% We hypothesized that the large-scale removal and/or silencing of TEs (whose proliferation may not have faced such strong purifying selection in wild populations) could be a major force in the domestication of strawberry.
% 
% In this study, utilizing a comparative genomics framework, we examined the role of transposable elements (TEs) in shaping the genome of cultivated strawberry.
% We hypothesized that TEs have played a role in the domestication of strawberry, and that the TE landscape around genes could be contributing to domestication-related phenotypes.
% Furthermore, given that 



% We draw comparisons between the wild diploid strawberry (\textit{Fragaria vesca} `H4', \textit{2n} $=$ 2X $=$ 14), the wild octoploid strawberry (\textit{Fragaria chiloensis} `Del Norte', \textit{2n} $=$ 8X $=$ 56), and the cultivated octoploid strawberry (\textit{Fragaria x ananassa} `Royal Royce', \textit{2n} $=$ 8X $=$ 56).




\section{Results:}
\subsection{TE Pangenomes Create Comparable Annotations and Show Differences in Whole-Genome TE Content:}
Our first objective was to create a consistent TE annotation for each genome, reducing false positives and negatives, by creating a TE pangenome.
The genomes for Fragaria x ananassa `Royal Royce', Fragaria chiloensis `Del Norte', and Fragaria vesca `H4' were used to generate a TE pangenome with EDTA (CITE).
This resulted in a genome repeat content of 38.28\% for Del Norte, 35.88\% for Royal Royce, and 28.62\% for H4.
Overall, the annotations for the octploid genomes are not all that different, differing by a total of only 2.2\%.
Table \ref{tab:genomes_combined_reduced} shows a summary table of the TE content for the three genomes, with the full version in the supplement (TODO).


% TODO THIS MUST BE CHECKED BECAUSE I fed my raw data into Chat jippity to format the LaTeX table. Numbers might be off
\begin{table}[ht]
\centering
\begin{tabularx}{\textwidth}{l|X|r|r|r}
\toprule
\textbf{Class} & \textbf{Subtype} & \textbf{H4 (\%masked)} & \textbf{DN (\%masked)} & \textbf{RR (\%masked)} \\
\toprule
\textbf{LINE} & L1 & 0.57\% & 0.46\% & 0.47\% \\
\textbf{LTR} & Copia & 3.46\% & 3.67\% & 3.51\% \\
& Gypsy & 5.05\% & 7.75\% & 7.90\% \\
& unknown & 7.62\% & 13.72\% & 12.57\% \\
\textbf{TIR} & CACTA & 3.71\% & 4.88\% & 3.86\% \\
& Mutator & 2.43\% & 2.35\% & 2.16\% \\
& PIF\_Harbinger & 0.97\% & 0.91\% & 0.91\% \\
& Tc1\_Mariner & 0.04\% & 0.02\% & 0.03\% \\
& hAT & 1.85\% & 2.16\% & 1.93\% \\
\textbf{nonTIR} & helitron & 1.87\% & 1.17\% & 1.24\% \\
\toprule
\textbf{Total Interspersed} & & \textbf{28.62\%} & \textbf{38.28\%} & \textbf{35.88\%} \\
\bottomrule
\end{tabularx}
\caption{Repeat elements in genomes H4, DN, and RR (\% masked) for LINE, LTR, and TIR and nonTIR categories. Other categories such as SINEs and `repeat fragment' are not shown for brevity.}
\label{tab:genomes_combined_reduced}
\end{table}
% The Del Norte genome is slightly larger... (TODO)

\subsection{Domesticated Strawberry is More Gene-Dense and Less TE-Dense}
We hypothesized that the TE content variation around genes could be contributing to domestication-related phenotypes.
In order to investigate the relationship between TE content and phenotypic variation in strawberry, we calculated the TE Density of genes in both genomes.
However, prior to calculating TE Density we determined the average distance between each in order to better inform the choice of window parameters.
We found that Royal Royce has a mean distance ($\mu$) and standard deviation ($\sigma$) of 5.01 and 8.08 KB; Del Norte was greater in both respects: $\mu = 5.20$ KB, $\sigma = 8.64$ KB.
Thus, we decided to utilize 5KB as the primary reference window for the following TE Density analyses.
We generated a histogram of the gene distances for both genomes, available in the supplement (TODO).

% TODO give this a better title or no title at all?
\subsubsection{Dotplots of TE Density Show General Trends of TE Presence:}
To get a sense of the general trends of gene-centric TE content, we first decided to plot the aggregate of our data: the average TE Density of all genes in Royal Royce and Del Norte as a function of measurement window size and location relative to genes (Figures \ref{fig:RR_order_dotplot} and \ref{fig:DN_order_dotplot}).
These graphs largely recapitulated the results from the whole-genome annotation, as we did not see TE Density differences greater than $\sim$ 2.0 - 2.5\% between Del Norte and Royal Royce in any one TE category, window, or distance combination.
Thus, in the aggregate, we do not see large TE differences between the genomes, and our TE Density results are parallel to the trends we see in the whole-genome annotations (Table \ref{tab:genomes_combined_reduced}).


\begin{figure}[ht]
\centering
\includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{../../results/dotplot/RR_Dotplot_Strawberry_AllGenes_Order.png}
\caption{Dotplot of average TE Density, delineated by TE Order classification, of all Royal Royce genes as a function of measurement window size and location relative to genes.}
\label{fig:RR_order_dotplot}
\end{figure}

\begin{figure}[ht]
\centering
\includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{../../results/dotplot/DN_Dotplot_Strawberry_AllGenes_Order.png}
\caption{Dotplot of average TE Density, delineated by TE Order classification, of all Del Norte genes as a function of measurement window size and location relative to genes.}
\label{fig:DN_order_dotplot}
\end{figure}

%!!!!!
\subsection{Few Functional Enrichments are Shared Between TE-Dense Genes of Wild and Domesticated Strawberry:} \label{sec:shared_unshared_go}
In order to further explore the subtle TE differences reported in our TE pangenome annotations and TE Density dotplots we examined the functional enrichments of the TE-dense genes.
We hypothesized that in the absence of differential selection on TE-gene relationships, the set of TE-dense genes from both genomes would share a large proportion of functional enrichments (GO terms).
However, we observed that the sets from both genomes were enriched for very different GO terms, and had little in common (Figure \ref{fig:go_compare_total_5K}).
Here, we excluded genes with 0 TE Density, and examined the 95th percentile of total TE Density for the 5 KB upstream window of all genes in Royal Royce and Del Norte.
The cutoff value for the 95th percentile was 0.872 and 0.901 respectively, which corresponds to 4360 and 4505 BP TE occupation of those windows.
Applying this cutoff to our dataset yielded 2507 and 2501 genes (NOTE scott cutoff tables).
After identifying these ``super-dense'' genes, we then subsetted our ortholog table to identify their \textit{Arabidopsis} orthologs and perform a GO enrichment analysis.
Supplemental Table \ref{tab:upset_numbers} shows how the number of genes changes as the analysis progressed.

We generated an UpSet plot to display the shared and unique GO terms associated with TE-dense genes of wild and domesticated strawberry (Figure \ref{fig:go_compare_total_5K}).
Despite the roughly equal number of genes in both datasets, the majority of the GO terms were unique to Del Norte, which had 160 unique terms compared to 83, and only 30 were shared with Royal Royce.
Of the 30 shared terms, GO:0006968 (cellular defense response), GO:0006970 (response to osmotic stress), GO:0009408 (response to heat) and GO:0009686 (gibberellin biosynthetic process) were the most notable; the full list of shared terms is available in the supplement (TODO supplement).

The domesticated strawberry unique terms included GO:00006002 (fructose 6-phosphate metabolic process), GO:0009413 (response to flooding), GO:0009751 (response to salicylic acid), GO:0080141 (regulation of jasmonic acid biosynthetic process), GO:1902265 (abscisic acid homeostasis), and GO:00:10188 (response to microbial phytotoxin).
The wild strawberry unique terms included GO:0010401 (pectic galactan metabolic process), GO:0009252 (peptidoglycan biosynthetic process, GO:0009835 (fruit ripening), and GO:0000024 (maltose biosynthetic process), 
among others.
A complete list of both sets is available in the supplement (TODO supplement).






\subsection{Domesticated Strawberry Has Less TEs Near Its Syntelogs With Wild Strawberry:}
Given the results from our whole genome annotations, we expected that the syntelog pairs would have similar levels of TE presence, but referencing Figure \ref{fig:syntelog_5000_total} we see rather large differences, and show that the Del Norte syntelog is on average more TE-dense.
% Here we plotted the difference in TE Density values of syntelog pairs, subtracting the Royal Royce value from the Del Norte value.
% The TE Density values reported in Figure \ref{fig:syntelog_5000_total} were derived from a 5 KB measurement window upstream of genes, and display the total TE Density --- the sum of all TE occupied base-pairs for the window.
Before we began statistically investigating this relationship, we first plotted the quantile-quantile (QQ) plot to assess the normality of the distribution.
The QQ plot showed that the data likely does not come from a normal distribution, and suggests that the data contains more extreme values than a normal distribution (TODO supplement).

Thus, we decided to move forward with a nonparametric test.
% TODO CHECK THIS STATEMENT ON WHAT EXACTLY A WILCOXIN IS IN CASE A STATS PERSON TRIES TO GRILL ME
We performed a Wilcoxon signed-rank test to test the null hypothesis ($H_{0}$) that TE Density values of the syntelog pairs come from the same distribution, and set a one-sided alternative hypothesis ($H_{a}$) positing that the Del Norte values are consistently greater than Royal Royce.
This test was performed on the 5KB upstream window, considering all TEs.
The mean TE Density difference was calculated as 0.032, which translates to a 161.6 BP bias in total TE content towards Del Norte.
Our resulting p-value of 2.730e-164 allows us to reject the null hypothesis in favor of the alternative.

We repeated this analysis for all TE types, over all windows, for both upstream and downstream, and found that there is no combination in which the Royal Royce syntelogs are more TE-dense (TODO supplement).
% TODO remove this sentence?
We also observed that the mean is always positive and that the absolute value of the 95th percentile cutoffs are consistently greater for the Del Norte-biased portion of the graphs.
Table \ref{tab:syntelog_summary} shows a summary of the syntelog TE Density differences for additional TE categories and measurement windows.
Of particular interest were CACTA elements, a superfamily of TIR elements, as they had some of the largest syntelog TE Density differences.
This category also displayed one of the largest differences in whole genome annotation between Del Norte and Royal Royce, 1.02\%, second only to unknown LTR elements (Table \ref{tab:genomes_combined_reduced}).
% It is worth noting that the generally large size of CACTA elements, which could impact their identification rate, proclivity to be near genes and or silenced and degraded, or other factors.

\begin{table}[ht]
\centering
\begin{tabularx}{\textwidth}{l|X|r|r}
\toprule
\textbf{TE Category} & \textbf{Window} & \textbf{BP Bias} & \textbf{P-Value} \\
\toprule
\textbf{Total TE} & 2500 & 113.7 & 1.603e-228 \\
& 5000 & 161.6 & 2.730e-164 \\
& 7500 & 202.3 & 4.871e-137 \\
\hline
\textbf{LTR} & 2500 & 125.6 & 5.776e-141 \\
& 5000 & 153.9 & 2.973e-99 \\
& 7500 & 171.9 & 7.172e-79 \\
\hline
\textbf{TIR} & 2500 & 73.2 & 4.140e-51 \\
& 5000 & 99.7 & 1.153e-41 \\
& 7500 & 126.7 & 5.753e-49 \\
\textbf{TIR/CACTA} & 2500 & 148.7 & 3.844e-33 \\
& 5000 & 218.0 & 1.813e-41 \\
& 7500 & 250.0 & 1.368e-48 \\
\bottomrule
\end{tabularx}
\caption{Summary Table of the TE Density differences between Del Norte and Royal Royce syntelogs. The BP Bias column reflects the average base-pair bias towards Del Norte, and the P-Value column reflects the significance of the Wilcoxon signed-rank test.}
\label{tab:syntelog_summary}
\end{table}


\begin{figure}[ht]
\centering
\includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{../../results/density_analysis/figures/Total_TE_Density_5000_Upstream_DensityDifferences.png}
\caption{
Histogram of non-zero differences in TE Density values of syntelogs between Del Norte and Royal Royce.
Royal Royce TE Density values were subtracted from Del Norte values; negative values reflect a higher TE value for the Royal Royce syntelog, and positive values reflect a higher TE value for the Del Norte syntelog.
Values were binned into groupings reflecting 5\% increases or decreases in TE Density.
All bins except for most positive bin are half-open.
For example, the leftmost bin reflects an interval of $[-1.0, -0.95)$, rightmost is $[0.95, 1.0]$.
% TODO REWORD THE FOLLOWING:
%The cutoff values for the 95th percentile of positive values, and the 5th percentile of negative values were calculated from the array of non-zero differences.
}
\label{fig:syntelog_5000_total}
\end{figure}

% TODO move this section?


% NOTE
% This is taken from Hardigan 2021
% """
% We found that genes targeted by selection were more likely to affect fruit development, cell wall metabolism, the regulation of gene expression, and the regulation and coordination of hormone signaling pathways, including auxin, abscisic acid, and gibberellic acid pathways. The latter have been shown to regulate expansion and ripening in nonclimacteric fruit
% """
\subsection{Domesticated Strawberry Has Less TE Density Around Flowering, Fruit and Sugar Related Syntelogs:}
Next we looked at the TE Density of the syntelogs that had the greatest differences in TE Density --- the tails of Figure \ref{fig:syntelog_5000_total}.
Here we examined only the genes that had a TE Density difference greater than or equal to $\lvert 0.75 \rvert$.
We then further subsetted our genes so that they had Arabidopsis orthologs and were in selective sweeps identified in Whitaker et al 2024 (TODO CITE https://academic.oup.com/plcell/article/36/5/1622/7479895).
Figure (TODO) shows the gene ``Fxa6Bg03714'', annotated as GO:0006002 (fructose 6-phosphate metabolic process), and its syntenic neighbors as well as a variety of TE insertions in the area.
It has a TE Density difference of $\lvert 0.960 \rvert$ (4800 BP of TE present in domesticated) and is located within a selective sweep (UCD Chr 6B 141) (TODO CITE).
The TE Density difference is not due to a multitude of TEs but rather a single large TE insertion.
There is an intact LTR/Gypsy element immediately upstream of Fxa6Bg03714.
This specific element is 12,810 BP in length and has an LTR identity of 1.00 indicating that it is a relatively young element and/or there is selection pressure to maintain the sequence.
The TE family of this element, TE\_00003582 is unique to the domesticated genome, and is not present in the wild genome.
There are 9 other instances of this family within the domesticated genome, further indicating that this is a relatively new TE in the genome.


Finally, we also observed that there is a gene, Fxa6Bg03713, overlapping and contained within the bounds of the Gypsy element.
This ``gene'' is not syntenic or homologous to anything within the wild genome, and a BLASTX search of its sequence against the NCBI database returned an unknown protein, suggesting that this other ``gene'' is a misannotation of an expressed TE rather than a gene sequence embedded within a TE (TODO check, CITE).

% \usepackage{multirow}
% TODO add cutoff value for 95th percentile to this?
% TODO this will probably be a supplemental table so we should renumber it
\begin{table}[]
\centering
\small
\begin{tabular}{l|l|l|l|l|l}
\toprule
\textbf{Genome} & \textbf{\# TE Dense Genes} & \textbf{\# Strawberry Gene w/ AT Orth} & \textbf{\# Unique AT Orth} & \textbf{\# Unique GO Terms} & \textbf{\# Shared GO Terms} \\
\bottomrule
\textbf{Royal Royce}     & 2507                        & 682                                        & 599                      & 113                   & \multirow{2}{*}{30}          \\
\textbf{Del Norte}       & 2501                        & 691                                        & 633                      & 190                   &                              \\
\bottomrule
\end{tabular}
\caption{\textit{NOTE candidate for supplement} Summary of the reduction in dataset size as we move from the 95th percentile of TE-dense genes for the upstream 5KB window to the final GO enrichment analysis.}
\label{tab:upset_numbers}
\end{table}

\begin{figure}[ht]
\centering
\includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{../../results/go_analysis/enrichment/upset_plots/nonsyntenic/UpSet_Total_TE_Density_5000_Upstream_NonSyntelog.png}
	\caption{UpSet plot of the GO term enrichments of the TE-dense genes of Royal Royce and Del Norte. This is a plot of the 95th percentile of TE-dense genes for the 5KB upstream window. The bottom bars show the number of shared GO terms, e.g there are a total of 113 GO terms in the Royal Royce dataset, 30 of which are shared with Del Norte and 83 are unique.}
	\label{fig:go_compare_total_5K}
\end{figure}


% \begin{table}[h]
% \begin{tabular}{lcccc}
% \textbf{Genome}                  & \multicolumn{1}{l}{\textbf{Direction}} & \multicolumn{1}{l}{\textbf{Cutoff Value}} & \multicolumn{1}{l}{\textbf{Genes Meeting Cutoff}} & \multicolumn{1}{l}{\textbf{Genes w/ AT Ortholog \& GO Term}} \\
% \multicolumn{1}{l|}{Royal Royce} & \multirow{2}{*}{Upstream}              & 0.607                                     & 1593                                              & 635                                                          \\
% \multicolumn{1}{l|}{Del Norte}   &                                        & 0.633                                     & 1380                                              & 448                                                          \\ \hline
% \multicolumn{1}{l|}{Royal Royce} & \multirow{2}{*}{Downstream}            & 0.617                                     & 1672                                              & 579                                                          \\
% \multicolumn{1}{l|}{Del Norte}   &                                        & 0.640                                     & 1370                                              & 588                                                         
% \end{tabular}
% \caption{Test}
% \label{tab:go_compare_TIR_5K}
% \end{table}

% TODO talk about the figures.
% \begin{figure}[ht]
% \centering
% \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{Figures/TIR_5000_Upstream_NonSyntelog.png}
% \caption{Test}
% \label{fig:new}
% \end{figure}




%\subsection{GO Annotations of the Syntelogs Greatly Differing in TE-Density Suggest that... }
%Next, we investigated the tails of Figure \ref{fig:syntelog_2500_total}, and its sister datasets to examine what kinds of functional enrichments exist for the syntelogs that greatly differ in TE density.
%We asked ourselves, ``What kinds of genes would have high TE density in the domesticated genome but hardly any in the wild genome?'' and vice-versa.
%First

%\subsection{Gene Comparisons TODO RETITLE:}
%\paragraph{Domesticated and Wild Strawberry Share Few GO Terms Among Their TE-Dense genes:}
%Next, we performed a GO enrichment on the TE-dense genes of each genome, regardless of whether or not that gene had a corresponding syntelog or large TE difference between the genomes (TODO most do because we had to get AT orthologs).
%Here we sought to compare and contrast the GO terms of the TE-dense genes to see if there were gene ontologies that are unique or shared between the genomes.

\subsection{TE-Dense Genes Exhibit Reduced Rates of Ortholog Identification:}
Generating functional enrichments of genes of interest is a fundamental step in analyzing genomic data and drawing conclusions about their biological significance.
Here we show that the TE-dense genes are less likely to have orthologs with closely-related strawberry species, as well as \textit{Arabidopsis} orthologs, when compared to a random sampling (10,000 iterations) of genes.
Figure \ref{fig:ortholog_survival_rate_barplot} shows the wild diploid strawberry ortholog and \textit{Arabidopsis} ortholog identification rates for the 95th-percentile of TE-dense genes in domesticated strawberry, and a random sampling of genes of the same size.
The random sample (n = 5,065 genes) identified an \textit{Arabidopsis} for 2,986 (58.96\%) of the genes, while the TE-dense genes (n = 5,065 genes) identified an \textit{Arabidopsis} ortholog for 1,801 (35.56\%) of the genes.
This ortholog identification rate for the TE-dense genes was -34.386 standard deviations away from the mean of the random sample (TODO supplement).
We repeated this analysis for the LTR and TIR datasets and had similar results; although the 95th percentiles for LTR and TIR were lower, and the ortholog identification rates were slightly higher.


Out of a concern that a greater portion of TE-dense genes may actually be TEs annotated as genes (false positives) and/or poorly supported or noisy gene models, and thus biased towards not having an ortholog, we filtered both the random sample and TE-dense genes to only include those with an AED score of less than or equal to 0.75.
This did not remove many genes from the dataset, and the ortholog identification rates remained consistent with the unfiltered dataset.
Histograms of the AED scores for all genes for all genomes are available in the supplement (TODO supplement).

\begin{figure}[ht]
\centering
\includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{../../results/density_analysis/cutoff_tables/ortholog_analysis/RR_Total_TE_Density_5000_Upstream_Upper_95_ortholog_barplot.png}
	\caption{Barplot of the ortholog identification rates for the 95th percentile of TE-dense genes in domesticated strawberry (Royal Royce) and a random sampling of genes of the same size. The ortholog identification rates are shown for the wild diploid (H4) strawberry and \textit{Arabidopsis}. Genes in both the random-sample and TE-dense set were filtered to exclude those with an AED score greater than 0.75. The 95th percentile cutoff evaluated to a TE Density value of 0.872 (4360 BP) for the upstream 5KB window.}
\label{fig:ortholog_survival_rate_barplot}
\end{figure}

\subsection{Deeply Conserved Genes Have Reduced TE Density:}
To complement our findings on the ortholog identification rates of TE-dense genes, we examined the TE Density of genes with \textit{Arabidopsis} orthologs.
At a broad-scale, we found that the TE Density of genes with \textit{Arabidopsis} orthologs was consistently lower than the TE Density of genes without \textit{Arabidopsis} orthologs.
Supplemental Figure (TODO supplement) shows the general pattern of TE Density of all genes, similar to Figure \ref{fig:RR_order_dotplot}, but each TE grouping is downshifted in their TE Density values.

We also performed a MannWhitneyU test.
Our null hypothesis ($H_{0}$) was that the two groups are equal in their TE Density, and our alternative hypothesis ($H_{a}$) was that the TE Density of genes with \textit{Arabidopsis} orthologs is less than the TE Density of genes without \textit{Arabidopsis} orthologs.
Our resulting p-value was 0 (TODO SCOTT rounding error in code?), the 







\section{Discussion:}
\paragraph{Interpretations of Functional Enrichment Analyses:}
The functional enrichments of the TE-dense genes of wild and domesticated strawberry presented in Figure \ref{fig:go_compare_total_5K} can be interpreted in a number of ways.
For example, TEs are broadly known to be negative fitness factors, and their presence near genes can lead to gene silencing and/or misregulation.
If the results are interpreted only in this light, then each functional enrichment would appear to be an instance of a TE negatively impacting the gene.
However TE insertions can be neutral or positive from the standpoint of expression.
TODO...
Additionally, the GO terms themselves are not always indicative of a gene's function as they are largely inferred from \textit{Arabidopsis} data.
Finally, the GO terms themselves may overlap in meaning given their hierarchical nature.
For example the GO term GO:0080141 (regulation of jasmonic acid biosynthetic process) was unique to the domesticated strawberry, while the GO term GO:0009695 (jasmonic acid biosynthetic process) was unique to the wild strawberry.


\paragraph{Limitations of our TE and Gene Annotations:}
Without extensive manual annotation, TE annotations generated from software are ultimately a hypothesis of the identification and classification of TEs in a genome.
The structural diversity of TEs, the quality of input genomes, the availability of reference data sets such as TE models and consensus sequences, the performance of \textit{de novo} annotation software, and the difficulty in manually verifying individual TE models comes together to limit the interpretation and tractability of TE annotations.
Additionally, not all TEs are created equal and some are more difficult to analyze than others.
For example, Helitron elements are notoriously difficult to analyze due to their structural assymetry, a lack of TSD upon integration and high sequence heterogeneity (TODO cite Helitron book chapter).

Similarly, gene annotations are also a hypothesis of the presence and structure of genes in a genome.
Both issues come together when we consider the TE Density of genes, as the TE Density of a gene is a function of the quality of the gene annotation and the TE annotation.

\paragraph{Shared GO Terms:}
In Section \ref{sec:shared_unshared_go} we showed that the TE-dense genes of the two genomes were enriched for very different GO terms, and had little in common.
We did not find the set of shared terms particularly interesting save for a few terms.
% However, GO:000940 (response to heat), was of particular interest due to previous research by Lisch et al (SCOTT CITE, go to twitter DM with Pat) in \textit{Zea mays}, and previous research that shows the activation of TEs under stress conditions (TODO).
% They showed that heat stress can reverse transcriptional silencing of TEs in DNA methylation deficient (RdDM specifically) mutants, as heat stress would reduce the amount of repressive histone modifications to the terminal inverted repeats of \textit{MuDr} elements.
% If there are genes with nearby TEs possessing aberrant or weakly imposed DNA methylation patterns, then it would make sense that this GO term showed up in both datasets.


\paragraph{Limitations of GO:}
The ability to infer gene function in strawberry is ultimately limited by our ability to infer orthology to \textit{Arabidopsis thaliana}.
Futhermore, not every \textit{Arabidopsis} gene is associated with a GO term.
Thus, it is entirely possible we are missing functional annotations more pertinent to strawberry domestication and agriculture, because the strawberry genes have diverged and/or acquired new functions not in the \textit{Arabidopsis} data set.


\section{Methods:} \label{sec:methods}
All code used to conduct this study is located within the GitHub repository at \url{https://github.com/sjteresi/Strawberry_Domestication}.
The Makefile, located within the root directory, is the primary reference for recreating the bioinformatic analyses.
Unless otherwise specified, analyses were conducted using Python v3.10.10.
Please see the requirements directory within the repository for a complete list of minor packages.
The TE Density and EDTA Git repositories were added as Git Submodules within the project repository to enable greater version control and reproducibility.

% \subsection{Plant Treatments:}
% Plants were grown with under a 16:8 photoperiod.
% Young leaf tissue was collected 5-6 hours after first light.
% Plants were subjected to 3 different temperature and humidity treatments.
% Plants were grown for a minimum of 14 days in each treatment prior to sampling.
% The warm treatment was set at 28 C and 80\% humidity.
% The neutral treatment was set at 20 C and 40\% humidity.
% The cold treatment was set at 12 C and 40\% humidity.
% New plants were used for each treatment.
% TODO but the neutral Del Norte that Maria has is being used for the cold treatment.

% \subsection{RNA Extraction:}
% TODO check with Maria
% 
% \subsection{RNA Sequencing:}
% TODO check with Maria

\subsection{TE Annotation:} \label{sec:EDTA_methods}
The genomes for Fragaria x ananassa `Royal Royce', Fragaria chiloensis `Del Norte', and Fragaria vesca `H4' were used to generate a TE pangenome with EDTA (CITE).
First, Royal Royce was annotated with EDTA v2.1.1 using the \verb|--sensitive 1| and \verb|--anno 1| flags to obtain a RepeatMasker output, due to issues with running EDTA v2.2.1 on Royal Royce.
Then each strawberry genome was individually annotated with EDTA v2.2.1, using the \verb|--sensitive 1| and \verb|--anno 1| flags; and the RepeatMasker output from the Royal Royce annotation was used as an additional optional input for that genome only.
Third, a pangenome repeat library was created by calling \verb|panEDTA.sh| on a directory containing the results from the individual annotations.
Finally, the pangenome library was used to re-annotate the individual genomes.

\subsection{TE Density Calculation:} \label{sec:density_methods}
TE Density (CITE) v2.2.1 was used to generate TE Density with default parameters using the EDTA-produced TE annotations.
Some modifications were made to the EDTA-produced TE annotations during the preprocessing stage of the TE Density pipeline; these modifications collapsed TE-family level groupings into the same TE superfamily, as consistent with the guidelines of the TE Density software.
These changes may be found within the \verb|import_strawberry_EDTA.py| script, available within the GitHub repository.





\subsection{Ortholog Identification:} \label{sec:methods_orthologs}
Genome-wide analyses using a combination of synteny- and reciprocal BLASTp- based approaches, were performed to identify orthologs amongst the strawberry genomes.
Ortholog pairs were excluded if they did not reside on homologous (TODO check term) chromosomes, i.e a gene pair from chromosome 1-4 in the Royal Royce genome was excluded from having a Del Norte ortholog on chromosome 2-3.
Data from Hardigan (CITE) was used to establish the chromosome relationships between Del Norte and Royal Royce.
Ortholog pairs with E-values greater than or equal to 0.05 were excluded from both datasets.
In the event that SynMap or BLASTp returned multiple genes for a single gene, the ortholog pair with the best (lowest) E-value score was kept, and the other pairs discarded.
In the event that a gene had an ortholog identified through both BLASTp and SynMap, we gave priority to the SynMap synteny results, preferentially keeping that ortholog pair. 

Orthologs between H4 and Arabidopsis, similarly identified using a combination of synteny and BLASTp approaches, were used to establish Arabidopsis orthology.
(TODO PAT, more detail here?)
GO terms were acquired from TAIR via the GOSLIM data set, versioned to March 1st, 2023 (cite).
TopGO v2.50.0 was used within R v4.2.2 to generate the GO enrichments.

\subsubsection{Single Copy Ortholog Identification:} \label{sec:methods_single_copy}
A list of genes retained as single copy orthologs (SCOs) over deep evolutionary time was acquired from \textit{De Smet et al.} \cite{DeSmet2013}.
These genes were then intersected with our ortholog table to identify the SCOs in the strawberry genomes.


\paragraph{Syntelog Identification with SynMap:} \label{sec:methods_syntelogs}
First, \href{https://genomevolution.org/CoGe/SynMap.pl}{SynMap} was run between the Del Norte and Royal Royce genome to identify syntelogs.
Then SynMap was run between the Royal Royce and H4 genomes.
In both cases SynMap was run with all default options, except for the fact that `Merge Syntenic Blocks` was set to `Quota Align Merge`.

The analysis of Royal Royce vs. Del Norte can be retrieved with the following \href{ https://genomevolution.org/r/1r9uk}{link}.
The analysis of Royal Royce vs. H4 can be retrieved with the following \href{https://genomevolution.org/r/1r9uy}{link}.
% TODO the genomes are private on CoGe currently, is this an issue?

\paragraph{Homolog Identification with BLASTp} \label{sec:methods_homologs}
BLAST v2.2.26 was used to identify homologs that may have been missed using a synteny-based approach.
Specifically, we used `blastall -p blastp` to perform the search. 

\subsubsection{KA KS Calculation:} \label{sec:ka_ks}
TODO

\section{Conclusion:}

\section{Declarations:}
\subsection{Ethics approval and consent to participate:}
Not applicable

\subsection{Consent for publication:}
Not applicable

\subsection{Availability of data and materials:}
All code used to conduct this study is located within the GitHub repository at \url{https://github.com/sjteresi/Strawberry_Domestication}.
The genome annotations, repeat library, TE Density data, and ortholog table is located on Dryad at TODO.

\subsection{Competing Interests:}
The authors declare that they have no competing interests

\subsection{Funding:}

\subsection{Authors' Contributions:}
S.J.T. and P.P.E. conceived and designed the project.
S.J.T wrote the code and performed analyses.
J.R.B provided guidance and helped generate figures.
Ning Jiang provided TE guidance and feedback.
S.J.T. wrote the manuscript draft, and all authors reviewed and revised the manuscripts.

\subsection{Acknowlegements:}
We thank Michael Teresi for software optimization advice, Jordan Brock for feedback on the manuscript, Adrian Platts for TE visualization and dataset acquisition, and Nicholas Panchy for computing cluster help. We also thank Mackenzie Jacobs and Maria Magallanes-Lundback for their help with lab work that was intended for, but not used, in this study.

\newpage
% TODO toggle this when moving to overleaf and get rid of the bibliography stuff at the beginning of the document
%\printbibliography
\bibliography{references}

\end{document}
